Gateway/Request/AdditionalDataLevel23DataBuilder.php

```diff
@@ -0,0 +1,118 @@
+ *                               ############
+ *
+ * Adyen Payment Module
+ *
+ * Copyright (c) 2022 Adyen N.V.
+ * This file is open source and available under the MIT license.
+ * See the LICENSE file for more info.
+ *
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Gateway\Request;
+
+use Adyen\Payment\Helper\ChargedCurrency;
+use Adyen\Payment\Helper\Config;
+use Adyen\Payment\Helper\Data;
+use Adyen\Payment\Helper\Requests;
+use Adyen\Util\Uuid;
+use Magento\Payment\Gateway\Helper\SubjectReader;
+use Magento\Payment\Gateway\Request\BuilderInterface;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\Order\Item;
+use Magento\Store\Model\StoreManagerInterface;
+
+class AdditionalDataLevel23DataBuilder implements BuilderInterface
+{
+    /**
+     * @var Config
+     */
+    private $config;
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+    /**
+     * @var Data
+     */
+    private $adyenHelper;
+    /**
+     * @var ChargedCurrency
+     */
+    private $chargedCurrency;
+    /**
+     * @var Requests
+     */
+    private $adyenRequestHelper;
+
+    public function __construct(
+        Config $config,
+        StoreManagerInterface $storeManager,
+        ChargedCurrency $chargedCurrency,
+        Requests $adyenRequestHelper,
+        Data $adyenHelper
+    )
+    {
+        $this->config = $config;
+        $this->storeManager = $storeManager;
+        $this->adyenHelper = $adyenHelper;
+        $this->chargedCurrency = $chargedCurrency;
+        $this->adyenRequestHelper = $adyenRequestHelper;
+    }
+
+    public function build(array $buildSubject)
+    {
+        $requestBody = [];
+        if ($this->config->sendLevel23AdditionalData($this->storeManager->getStore()->getId())) {
+            $paymentDataObject = SubjectReader::readPayment($buildSubject);
+            $payment = $paymentDataObject->getPayment();
+            /** @var Order $order */
+            $order = $payment->getOrder();
+            $currencyCode = $this->chargedCurrency->getOrderAmountCurrency($order)->getCurrencyCode();
+
+            $prefix = 'enhancedSchemeData';
+            $requestBody['additionalData'][$prefix . '.totalTaxAmount'] = $this->adyenHelper->formatAmount($order->getTaxAmount(), $currencyCode);
+            $requestBody['additionalData'][$prefix . '.customerReference'] = $this->adyenRequestHelper->getShopperReference($order->getCustomerId(), $order->getIncrementId());
+            if ($order->getIsNotVirtual()) {
+                $requestBody['additionalData'][$prefix . '.freightAmount'] = $this->adyenHelper->formatAmount($order->getBaseShippingAmount(), $currencyCode);
+                $requestBody['additionalData'][$prefix . '.destinationPostalCode'] = $order->getShippingAddress()->getPostcode();
+                $requestBody['additionalData'][$prefix . '.destinationCountryCode'] = $order->getShippingAddress()->getCountryId();
+            }
```

Comment:
We can build `additionalData` or `riskdata` in a bit shorter way:
 
```suggestion
            $prefix = 'enhancedSchemeData.';
            $customerReference = $this->adyenRequestHelper->getShopperReference($order->getCustomerId(), $order->getIncrementId());
            $requestBody['additionalData'] = [
                $prefix . 'totalTaxAmount' => $this->adyenHelper->formatAmount($order->getTaxAmount(), $currencyCode),
                $prefix . 'customerReference' => $customerReference
            ];
            if ($order->getIsNotVirtual()) {
                $requestBody['additionalData'] += [
                    $prefix . 'freightAmount' => $this->adyenHelper->formatAmount($order->getBaseShippingAmount(), $currencyCode),
                    $prefix . 'destinationPostalCode' => $order->getShippingAddress()->getPostcode(),
                    $prefix . 'destinationCountryCode' => $order->getShippingAddress()->getCountryId(),
                ];
            }
```

Reply:
Nice suggestion @michaelpaul but since it's only a slight syntactical difference I wouldn't want to do another round of reviews over it. Maybe in a subsequent change we could add this